### PR TITLE
#1939 bug when clearing inferred native store

### PR DESCRIPTION
**Pull request by hmottestad**
Test to show failure.

### CODE REVIEW

**Comment by jeenbroekstra** [comment]
The test does not fail on RDFSchemaNativeRepositoryConnectionTest, which uses the deprecated ForwardChainingInferencer. On that basis I'd suspect the caching mechanism in the SchemaCachingInferencer somehow not being fully reset. It's weird that it doesn't fail on the memorystore though. Perhaps a discrepancy in how the notification mechanism works?

**Comment by jeenbroekstra** [comment]
I've just noticed that if we revert the recent optimization fix from GH-1932, this test no longer fails.

**Comment by jeenbroekstra** [comment]
What is interesting is that the extra statement is an inferred statement. If you adapt the test to only retrieve explicit statements at the end, the result is zero. So for some reason, the explicitly added statement is removed, but an inferred "copy" remains.

**Comment by hmottestad** [comment]
I didn't look much into it. Just realised it when working on the snapshot isolation branch. It's great you've found out that it's only for the SchemaCaching... that should narrow down a fix. This issue only seems to apply when adding a statement that is otherwise inferred, so I wouldn't call it critical.

File: core/sail/base/src/main/java/org/eclipse/rdf4j/sail/inferencer/InferencerConnectionWrapper.java [hunk 1]
```diff
@@ -93,7 +98,12 @@ public void prepare() throws SailException {
 	@Override
 	public void commit() throws SailException {
 		flushUpdates();
+		long count = getWrappedConnection().getStatements(null, null, null, true).stream().distinct().count();
 		super.commit();
+		long count2 = getWrappedConnection().getStatements(null, null, null, true).stream().distinct().count();
+		if (count != count2) {
+			System.out.println(count + " != " + count2);
+		}
 	}
```
Comment (hmottestad) [inline core/sail/base/src/main/java/org/eclipse/rdf4j/sail/inferencer/InferencerConnectionWrapper.java]: What I've found out is that the count is correct before calling `super.commit()` but wrong after.

**Pull request closed without merging**